### DISCUSSION TITLE
Saving GDSII after repair_design

### DISCUSSION THREAD

**Discussion by yathAg** [The-OpenROAD-Project/OpenROAD-flow-scripts#847] [Q&A] [answered]
I ran the flow for the asap7 ibex design. After opening the OpenROAD GUI using `make gui_final`  and reporting the timing metrics with the following commands, I get.

```
report_worst_slack
report_tns
report_wns
```
```
worst slack -107.32
tns -870.89
wns -107.32

```
I further proceed to run 

```
estimate_parasitics -placement
repair_design
```

and then the reported timing metrics are
```
worst slack 6.37
tns 0.00
wns 0.00
```

I have a few questions related to this procedure 
1) How do I save the updated def/odb and sdc files? 
2) How do I generate the new GDS files?

**Comment by rovinski** [comment]
Your two timing reports are likely using different parasitics modeling. The former uses extracted parasitics from the actual detailed routing. The latter uses parasitics estimated only from placement. This is comparing apples to oranges.

For a valid comparison you need to rerun global routing, then detailed routing, then parasitic extraction.

> How do I save the updated def/odb and sdc files?

`write_def`, `write_db`, and `write_sdc`, respectively. Mind you, you will be writing out invalid files until you rerun the necessary steps.

> How do I generate the new GDS files?

Save the .def file and then use the conversion script in ORFS.

**Reply by yathAg** [reply]
Thank you for the explanation and answers to the queries.

For the steps, I was referring to [the guide here ](https://openroad-flow-scripts.readthedocs.io/en/latest/tutorials/FlowTutorial.html#timing-optimization-using-repair-design)

" The repair_design command inserts buffers on nets to repair max slew, max capacitance and max fanout violations and on long wires to reduce RC delay. It also resizes gates to normalize slews. Use estimate_parasitics -placement before repair_design to account for estimated post-routing parasitics." 

Here the last line mentions `post-routing parasitics`. Are there any steps I missed from the tutorial, or is there any information I am missing?

Also, according to the tutorial, it should fix the timing issues? Or `repair_timing` is the appropriate command for fixing them?

**Reply by yathAg** [reply]
with repair timing i get 

```
>>> report_worst_slack
... report_tns
... report_wns
worst slack -107.32
tns -870.89
wns -107.32
>>> repair_timing
[WARNING RSZ-0021] no estimated parasitics. Using wire load models.
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _19375_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _19315_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18920_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18920_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver wire357/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18920_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver wire357/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18920_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver wire357/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _20094_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18920_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver wire357/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _20094_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew361/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18920_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver wire357/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _20094_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew361/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18920_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver wire357/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _20094_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew361/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18416_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _20155_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _33735_/SN
[WARNING RSZ-0075] makeBufferedNet failed for driver _18936_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _24359_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _36313_/QN
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew290/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _18920_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver wire357/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver _20094_/Y
[WARNING RSZ-0075] makeBufferedNet failed for driver load_slew361/Y
[INFO RSZ-0040] Inserted 14 buffers.
[INFO RSZ-0041] Resized 4 instances.
[INFO RSZ-0033] No hold violations found.
>>> report_worst_slack
... report_tns
... report_wns
worst slack 0.78
tns 0.00
wns 0.00
```

So with this, I should be able to save the def and sdc files and generate a new gdsii file? Or should I follow the steps you described in your previous reply?

**Reply by vijayank88** [reply]
> Thank you for the explanation and answers to the queries.
> 
> For the steps, I was referring to [the guide here ](https://openroad-flow-scripts.readthedocs.io/en/latest/tutorials/FlowTutorial.html#timing-optimization-using-repair-design)
> 
> " The repair_design command inserts buffers on nets to repair max slew, max capacitance and max fanout violations and on long wires to reduce RC delay. It also resizes gates to normalize slews. Use estimate_parasitics -placement before repair_design to account for estimated post-routing parasitics."
> 
> Here the last line mentions `post-routing parasitics`. Are there any steps I missed from the tutorial, or is there any information I am missing?
> 
> Also, according to the tutorial, it should fix the timing issues? Or `repair_timing` is the appropriate command for fixing them?

`post-routing parasitic` it is typo mistake. It is `placement parasitic`.

**Reply by maliberty** [reply]
@vijayank88 please make a PR to fix this and the missing variable below.

**Comment by maliberty** [comment]
I don't think you should try to run repair design at the end of the flow.  That's not a flow we support at this point.  You could try increasing SETUP_SLACK_MARGIN to overfix earlier in the flow.

**Reply by yathAg** [reply]
This would have to be set in the `./flow/designsasap7/<ibex>/config.mk` file right? 

Is there a document where all the variables for the config file are described or a way I could get a better understanding of these variables?

Thank you for the help!

**Reply by vijayank88** [reply]
https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/852 updated flow variables now.

**Comment by maliberty** [comment, marked as answer]
Yes.

Many are documented in https://openroad-flow-scripts.readthedocs.io/en/latest/user/FlowVariables.html though this one looks to be missing.  

The ultimate truth is in https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/tree/master/flow/scripts where the actual scripts that use the variables live.

**Reply by yathAg** [reply]
Okay. Thanks for the information!
